src/v/kafka/server/tests/quota_manager_test.cc

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "kafka/server/quota_manager.h"
+#include "random/generators.h"
+#include "resource_mgmt/rate.h"
+
+#include <seastar/core/lowres_clock.hh>
+
+#include <boost/test/tools/old/interface.hpp>
+#include <boost/test/unit_test.hpp>
+
+#include <chrono>
+
+using namespace std::chrono_literals;
+
+void set_quota_manager_test_defaults(
+  int16_t num_windows,
+  std::chrono::milliseconds window_size,
+  uint32_t quota_rate,
+  std::chrono::milliseconds max_delay) {
+    /// Configure quota manager with known defaults otherwise if assumptions are
+    /// made on tests defaults values obtained from config settings, test would
+    /// fail if these defaults are later changed
+    auto& cfg = config::shard_local_cfg();
+    cfg.get("default_num_windows").set_value(num_windows);
+    cfg.get("default_window_sec").set_value(window_size);
+    cfg.get("target_quota_time_rate")
+      .set_value(std::optional<uint32_t>(quota_rate));
+    cfg.get("max_kafka_throttle_delay_ms").set_value(max_delay);
+}
+
+/// This test emmulates a single client making frequent requests to
```

Comment:
Would it make sense to add a test with several clients, each with short requests, and ensuring they don't get throttled? Just to get some coverage for multi-client scenarios

Reply:
Yeah i actually have been thinking about testing, this simple unit test just ensures the math is sane. Its hard to test these types of features with real world clients where a non known amount of time is to be taken for each request made because up-front assertions cannot be performed. Knowing they don't get throttled is an OK test but i'd like to do better by actually verifying the returned `throttle_time_ms` value with an expected.